db/migrate/20150716133712_add_unique_index_on_journals.rb

```diff
@@ -0,0 +1,171 @@
+  end
+
+  def down
+    remove_index :journals, [:journable_type, :journable_id, :version]
+  end
+
+  private
+
+  def cleanup_duplicate_journals
+    duplicate_pairs = find_duplicate_journal_ids
+
+    if duplicate_pairs.any?
+      say "Found #{duplicate_pairs.count} journals with at least one duplicate!"
+      say_with_time 'Safely removing duplicates...' do
+        undeleted_pairs = []
+        duplicate_pairs.each do |current_id, duplicate_id|
+          say "Comparing journals ##{current_id} & ##{duplicate_id} for equality", subitem: true
+
+          current = MigrationHelperJournal.find(current_id)
+          duplicate = MigrationHelperJournal.find(duplicate_id)
+
+          if journals_equivalent?(current, duplicate)
+            say "Deleting journal ##{current.id}...", subitem: true
+            current.destroy
+          else
+            undeleted_pairs << [current_id, duplicate_id]
+          end
+        end
+
+        abort_on_undeleted_pairs undeleted_pairs
+      end
+    end
+  end
+
+  def find_duplicate_journal_ids
+    this = MigrationHelperJournal.table_name
+    MigrationHelperJournal
+      .joins("INNER JOIN #{this} other
+      ON #{this}.journable_id = other.journable_id AND
+         #{this}.journable_type = other.journable_type AND
+         #{this}.version = other.version")
+      .where("#{this}.id < other.id")
+      .select("#{this}.id id, other.id duplicate_id")
+      .uniq_by(&:id)
+      .map { |pair| [pair.id, pair.duplicate_id] }
+  end
+
+  def journals_equivalent?(a, b)
+    unless records_equivalent?(a, b)
+      say 'Difference found in table journals', subitem: true
+      return false
+    end
+
+    unless records_equivalent?(a.data, b.data)
+      say 'Difference found in related data table (e.g. work_package_journals)', subitem: true
+      return false
+    end
+
+    a_attachments = a.attachable_journals.pluck(:attachment_id).sort
+    b_attachments = b.attachable_journals.pluck(:attachment_id).sort
+    unless a_attachments == b_attachments
+      say 'Difference found in attachable_journals', subitem: true
+      return false
+    end
+
+    a_custom_fields = customizable_journals_to_hash a.customizable_journals
+    b_custom_fields = customizable_journals_to_hash b.customizable_journals
+    unless a_custom_fields == b_custom_fields
+      say 'Difference found in customizable_journals', subitem: true
+      return false
+    end
+
+    true
+  end
+
+  def records_equivalent?(a, b)
+    if a.nil? || b.nil?
+      return a == b
+    end
+
+    ignored = [:id, :journal_id, :created_at, :updated_at]
+    a.attributes.symbolize_keys.except(*ignored) == b.attributes.symbolize_keys.except(*ignored)
+  end
+
+  def customizable_journals_to_hash(customizable_journals)
+    customizable_journals.inject({}) { |hash, custom_journal|
+      hash[custom_journal.custom_field_id] = custom_journal.value
+      hash
+    }
+  end
+
+  def abort_on_undeleted_pairs(undeleted_pairs)
```

Comment:
I took your input and tried a few variants, but none gave me a particularly better readable version.

I now extracted a short named method that always passes the `subitem` parameter, because that was rally bad for readability (IMO), however, this scores me no points on the ABC scale ^^